src/Storages/ExternalStream/Kafka/KafkaSource.cpp

```diff
@@ -44,6 +45,9 @@ KafkaSource::KafkaSource(
     , topic(topic_)
     , shard(shard_)
     , offset(offset_)
+    , high_watermark(high_watermark_)
+    /// if offset == high_watermark, it means there is no message to read, so it already reaches the end
+    , reached_the_end(high_watermark != Kafka::NO_WARTERMARK && offset == high_watermark)
```

Comment:
If we reach to end initially, can avoid creating such source in the first place ?

Reply:
There is already a special handling for this scenario, which is that, when low watermark == high water mark, we use `NullSource`, which is more light-weight, instead of `KafkaSource`. Here, we still have this check in KafkaSource is just to let `KafkaSource` to handle this situation properly, so that logic inside `KafkaSource` is complete.